README.md
# NightView

[![Build Status](https://travis-ci.org/Boris-Em/NightView.svg?branch=master)](https://travis-ci.org/Boris-Em/NightView)
[![Version](https://img.shields.io/cocoapods/v/NightView.svg?style=flat)](http://cocoadocs.org/docsets/NightView)
[![License](https://img.shields.io/cocoapods/l/NightView.svg?style=flat)](http://cocoadocs.org/docsets/NightView)
[![Platform](https://img.shields.io/cocoapods/p/NightView.svg?style=flat)](http://cocoadocs.org/docsets/NightView)

<p align="center"><img src="https://s32.postimg.org/6w9jyums5/Night_View.jpg"/></p>	

**NightView** is an open source library that generates beautiful, random, starry skies on iOS.  

## Table of Contents

* [**Project Details**](#project-details)  
  * [Requirements](#requirements)
  * [License](#license)
  * [Support](#support)
  * [Sample App](#sample-app)
* [**Getting Started**](#getting-started)
  * [Installation](#installation)
  * [Setup](#setup)
* [**Documentation**](#documentation)
 
  
  ## Project Details
Learn more about the **NightView** project, licensing, support etc.

<p align="center"><img src="https://s31.postimg.org/tn5o1yikb/Night_View_i_Phone.jpg"/></p>

### Requirements
 - Requires iOS 7 or later. The sample project is optimized for iOS 9.
 - Requires Automatic Reference Counting (ARC).
 - Optimized for ARM64 Architecture.

### License
See the [License](https://github.com/Boris-Em/NightView/blob/master/LICENSE). You are free to make changes and use this in either personal or commercial projects. Attribution is not required, but highly appreciated. A little "Thanks!" (or something to that affect) is always welcome. If you use **NightView** in your app, please let us know!

### Support
[![https://gitter.im/Boris-Em/NightView](https://badges.gitter.im/Boris-Em/NightView.svg)](https://gitter.im/Boris-Em/NightView?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)  

Join us on gitter if you have any question!

### Sample App
The iOS Sample App included with this project demonstrates one way to correctly setup and use **NightView**. It also offers the possibility to customize the view within the app.

## Getting Started
It only takes a few simple steps to install and setup **NightView** to your project.

###Installation
The easiest way to install **NightView** is to use <a href="http://cocoapods.org/" target="_blank">CocoaPods</a>. To do so, simply add the following line to your `Podfile`:
	<pre><code>pod 'NightView'</code></pre>
	
The other way to install **NightView**, is to drag and drop the *NightView* folder into your Xcode project. When you do so, make sure to check the "*Copy items into destination group's folder*" box.

### Setup
**NightView** is a simple UIView subclass. It can be initialized with Interface Builder, or programatically.
 
 **Interface Builder Initialization**  
 1 - Drag a `UIView` to your `UIViewController`.  
 2 - Change the class of the new `UIView` to `NightView`.  
 3 - Select the `NightView` and open the Attributes Inspector. Most of the customizable properties can easily be set from the Attributes Inspector. The Sample App demonstrates this capability.
 
 **Programmatical Initialization**  
 Here is an example illustrating how to initialize a NightView instance programmatically:

 ```swift
 let nightView = NightView(frame: CGRectMake(0.0, 0.0, 200.0, 200.0))
 view.addSubview(nightView)
 ```
 
## Documentation
All of the methods and properties available for **NightView** are documented below.

##### `numberOfPointsForStar`  
The number of points for each star. For example, setting this property to 1, means that there will be 1 star for every point in the view. A greater number means less stars within the view. *Defaults to 10000.0.*

##### `starSize`  
The size of the stars in points. Note that setting `starSizeMinRandomizer` and `starSizeMaxRandomizer` will also influence the size of stars. *Defaults to 5.0.*

##### `starColor`  
The color of the stars. *Defaults to white.*

##### `starSizeMinRandomizer`  
The size of the stars is randomized within the `starSize` value. This property sets the minimum percent by which the stars' size could be changed. For example, a value of 50.0 means that the minimum size of a star will be 50% of the `starSize` property. *Defaults to 50.0.*

##### `starSizeMaxRandomizer`  
The size of the stars is randomized within the `starSize` value. This property sets the maximum percent by which the stars' size could be changed. For example, a value of 150.0 means that the maximum size of a star will be 150% of the `starSize` property. *Defaults to 150.0.*  

##### `minStarOpacity`  
The stars are drawn with a smaller opacity at the bottom of the view than at the top. This property sets the minimum opacity for the lower stars. Note that the stars at the top of the view will always have an opacity of 1.0. *Defaults to 0.5.*

##### `glowingIntensity`  
The intensity of the glowing of the stars, from 0 to 1. If set to 0, the stars will not glow. *Defaults to 0.5.*

##### `glowingDuration`  
The duration in seconds at which the stars glow. *Defaults to 2.0.*

##### `starType`  
The type of stars to be drawn. *Defaults to `.Round`.*  
<p align="center"><img src="https://s32.postimg.org/5lqa389lh/Night_View_Stars.jpg"/>

##### `reload()`  
Reloads the **NightView** instance, redrawing all of the stars.


CHANGELOG.md
# Change Log

## [v1.1.0](https://github.com/Boris-Em/NightView/releases/tag/1.0.1) (2016-07-11)
[Full Changelog](https://github.com/Boris-Em/BEMCheckBox/compare/1.0.0...1.0.1)
  
- Fix starType enum.
- Fix bug where NightView wasn't available in different modules.
- Add title to Customization view controller.


## [v1.0.0](https://github.com/Boris-Em/NightView/releases/tag/1.0.0) (2016-07-10)
First stable release of **NightView**


NightViewSampleProject/NightViewSampleProject/AppDelegate.swift
//
//  AppDelegate.swift
//  NightViewSampleProject
//
//  Created by Bobo on 7/10/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



NightViewSampleProject/NightViewSampleProject/ViewController.swift
//
//  ViewController.swift
//  NightViewSampleProject
//
//  Created by Bobo on 7/10/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var nightView: NightView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "toCustomization" {
            if let navVC = segue.destinationViewController as? UINavigationController, let customizationVC = navVC.topViewController as? CustomizationTableViewController {
                customizationVC.nightView = nightView
            }
        }
    }

}



NightViewSampleProject/NightViewSampleProjectTests/NightViewTests.swift
//
//  NightViewTests.swift
//  NightViewSampleProject
//
//  Created by Bobo on 7/11/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import XCTest
@testable import NightViewSampleProject

class NightViewTests: XCTestCase {

    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testInitWithFrame() {
        let nightView = NightView(frame: CGRectMake(0.0, 0.0, 500.0, 500.0))
        XCTAssertNotNil(nightView, "A fully initialized NightView instance should be instanciated.")
    }
    
    func testReload() {
        let nightView = NightView(frame: CGRectMake(0.0, 0.0, 500.0, 500.0))
        nightView.reload()
        let stars = nightView.layer.sublayers?.first?.sublayers
        XCTAssert(stars?.count > 0, "Multiple stars should have been drawn")
        
        nightView.reload()
        let newStars = nightView.layer.sublayers?.first?.sublayers
        XCTAssertEqual(stars?.count, newStars?.count, "After reloading, the same amount of stars should have been re-drawn")
        XCTAssertNotEqual(stars!, newStars!)

    }
    
    func testNumberOfPointsForStar() {
        let nightView = NightView(frame: CGRectMake(0.0, 0.0, 100.0, 100.0))
        nightView.reload()
        var stars = nightView.layer.sublayers?.first?.sublayers
        XCTAssert(stars?.count > 0)
        
        nightView.numberOfPointsForStar = 1
        stars = nightView.layer.sublayers?.first?.sublayers
        XCTAssertEqual(CGFloat(stars!.count), nightView.frame.size.width * nightView.frame.size.height)
    }
    
    func testStarSize() {
        let starSize: CGFloat = 100.0
        
        let nightView = NightView(frame: CGRectMake(0.0, 0.0, 50.0, 50.0))
        nightView.numberOfPointsForStar = 1
        nightView.starSize = starSize
        nightView.reload()
        var stars = nightView.layer.sublayers?.first?.sublayers
        
        var didFindBigStar = false
        var didFindSmallStar = false
        for star in stars! {
            if star.frame.size.height > starSize {
                didFindBigStar = true
            } else if star.frame.size.height < starSize {
                didFindSmallStar = true
            }
        }
        XCTAssertTrue(didFindSmallStar)
        XCTAssertTrue(didFindBigStar)
        
        
        nightView.starSizeMinRandomizer = 120.0
        nightView.starSizeMaxRandomizer = 150.0
        stars = nightView.layer.sublayers?.first?.sublayers
        
        didFindBigStar = false
        didFindSmallStar = false
        for star in stars! {
            if star.frame.size.height > starSize {
                didFindBigStar = true
            } else if star.frame.size.height < starSize {
                didFindSmallStar = true
            }
        }
        XCTAssertFalse(didFindSmallStar)
        XCTAssertTrue(didFindBigStar)

        
        nightView.starSizeMinRandomizer = 20.0
        nightView.starSizeMaxRandomizer = 80.0
        stars = nightView.layer.sublayers?.first?.sublayers
        
        didFindBigStar = false
        didFindSmallStar = false
        for star in stars! {
            if star.frame.size.height > starSize {
                didFindBigStar = true
            } else if star.frame.size.height < starSize {
                didFindSmallStar = true
            }
        }
        XCTAssertTrue(didFindSmallStar)
        XCTAssertFalse(didFindBigStar)

        
        nightView.starSizeMinRandomizer = 100.0
        nightView.starSizeMaxRandomizer = 100.0
        stars = nightView.layer.sublayers?.first?.sublayers
        
        didFindBigStar = false
        didFindSmallStar = false
        for star in stars! {
            if star.frame.size.height > starSize {
                didFindBigStar = true
            } else if star.frame.size.height < starSize {
                didFindSmallStar = true
            }
            XCTAssertEqual(star.frame.size.height, starSize)
        }
        XCTAssertFalse(didFindSmallStar)
        XCTAssertFalse(didFindBigStar)
    }
    
    func testMinStarOpacity() {
        let nightView = NightView(frame: CGRectMake(0.0, 0.0, 50.0, 50.0))
        nightView.numberOfPointsForStar = 1
        nightView.reload()
        var stars = nightView.layer.sublayers?.first?.sublayers
        
        var didFindLowOpacityStar = false
        for star in stars! {
            if star.opacity < 1.0 {
                didFindLowOpacityStar = true
                break
            }
        }
        XCTAssertTrue(didFindLowOpacityStar, "By default the stars at the bottom of the view should have a lower opacity")
        
        nightView.minStarOpacity = 1.0
        stars = nightView.layer.sublayers?.first?.sublayers
        
        didFindLowOpacityStar = false
        for star in stars! {
            if star.opacity < 1.0 {
                didFindLowOpacityStar = true
                break
            }
        }
        XCTAssertFalse(didFindLowOpacityStar, "All of the stars should have an opacity of 1")
    }

}


NightViewSampleProject/NightViewSampleProject/CustomizationTableViewController.swift
//
//  CustomizationTableViewController.swift
//  NightViewSampleProject
//
//  Created by Bobo on 7/11/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import UIKit

class CustomizationTableViewController: UITableViewController {

    @IBOutlet weak var numberOfPointsForStarLabel: UILabel!
    @IBOutlet weak var starSizeLabel: UILabel!
    @IBOutlet weak var starSizeMinRandomizerLabel: UILabel!
    @IBOutlet weak var starSizeMaxRandomizerLabel: UILabel!
    @IBOutlet weak var minStarOpacityLabel: UILabel!
    @IBOutlet weak var glowingIntensityLabel: UILabel!
    @IBOutlet weak var glowingDurationLabel: UILabel!
    @IBOutlet weak var starTypeLabel: UILabel!
    
    @IBOutlet weak var numberOfPointsForStarSlider: UISlider!
    @IBOutlet weak var starSizeSlider: UISlider!
    @IBOutlet weak var starSizeMinRandomizerSlider: UISlider!
    @IBOutlet weak var starSizeMaxRandomizerSlider: UISlider!
    @IBOutlet weak var minStarOpacitySlider: UISlider!
    @IBOutlet weak var glowingIntensitySlider: UISlider!
    @IBOutlet weak var glowingDurationSlider: UISlider!
    @IBOutlet weak var starTypeSegmentedControl: UISegmentedControl!
    
    var nightView: NightView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        title = "Customization"
        
        if let numberOfPointsForStar = nightView?.numberOfPointsForStar {
            numberOfPointsForStarLabel.text = String(format: "%.1f", numberOfPointsForStar)
            numberOfPointsForStarSlider.value = Float(numberOfPointsForStar)
        }
        
        if let starSize = nightView?.starSize {
            starSizeLabel.text = String(format: "%.1f", starSize)
            starSizeSlider.value = Float(starSize)
        }
        
        if let starSizeMinRandomizer = nightView?.starSizeMinRandomizer {
            starSizeMinRandomizerLabel.text = String(format: "%.1f", starSizeMinRandomizer)
            starSizeMinRandomizerSlider.value = Float(starSizeMinRandomizer)
        }
        
        if let starSizeMaxRandomizer = nightView?.starSizeMaxRandomizer {
            starSizeMaxRandomizerLabel.text = String(format: "%.1f", starSizeMaxRandomizer)
            starSizeMaxRandomizerSlider.value = Float(starSizeMaxRandomizer)
        }
        
        if let minStarOpacity = nightView?.minStarOpacity {
            minStarOpacityLabel.text = String(format: "%.1f", minStarOpacity)
            minStarOpacitySlider.value = Float(minStarOpacity)
        }
        
        if let glowingIntensity = nightView?.glowingIntensity {
            glowingIntensityLabel.text = String(format: "%.1f", glowingIntensity)
            glowingIntensitySlider.value = Float(glowingIntensity)
        }
        
        if let glowingDuration = nightView?.glowingDuration {
            glowingDurationLabel.text = String(format: "%.1f", glowingDuration)
            glowingDurationSlider.value = Float(glowingDuration)
        }
        
        if let starType = nightView?.starType {
            switch starType {
            case .Round:
                starTypeSegmentedControl.selectedSegmentIndex = 0
                break
            case .FiveBranchStar:
                starTypeSegmentedControl.selectedSegmentIndex = 1
                break
            }
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    @IBAction func handleSlideNumberOfPointsForStarSlider(sender: UISlider) {
        nightView?.numberOfPointsForStar = CGFloat(sender.value)
        numberOfPointsForStarLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleStarSizeSlider(sender: UISlider) {
        nightView?.starSize = CGFloat(sender.value)
        starSizeLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleStarSizeMinRandomizerSlider(sender: UISlider) {
        nightView?.starSizeMinRandomizer = Double(sender.value)
        starSizeMinRandomizerLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleStarSizeMaxRandomizerSlider(sender: UISlider) {
        nightView?.starSizeMaxRandomizer = Double(sender.value)
        starSizeMaxRandomizerLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleMinStarOpacitySlider(sender: UISlider) {
        nightView?.minStarOpacity = sender.value
        minStarOpacityLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleGlowingIntensitySlider(sender: UISlider) {
        nightView?.glowingIntensity = sender.value
        glowingIntensityLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleGlowingDurationSlider(sender: UISlider) {
        nightView?.glowingDuration = Double(sender.value)
        glowingDurationLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func handleTapDoneButton(sender: AnyObject) {
        dismissViewControllerAnimated(true, completion: nil)
    }
    
    @IBAction func handleStarTypeSegmentedControl(sender: UISegmentedControl) {
        switch sender.selectedSegmentIndex {
        case 0:
            nightView?.starType = .Round
            break
        default:
            nightView?.starType = .FiveBranchStar
            break
        }
    }
}


NightViewSampleProject/NightViewSampleProjectTests/NightViewSampleProjectTests.swift
//
//  NightViewSampleProjectTests.swift
//  NightViewSampleProjectTests
//
//  Created by Bobo on 7/10/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import XCTest
@testable import NightViewSampleProject

class NightViewSampleProjectTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock {
            // Put the code you want to measure the time of here.
        }
    }
    
}


NightView/NightView.swift
//
//  NightView.swift
//  Go
//
//  Created by Bobo on 7/8/16.
//  Copyright © 2016 Boris Emorine. All rights reserved.
//

import UIKit

/** Dazzling Nights on iOS.
*/
@IBDesignable
public class NightView: UIView {
    
    /** The number of points for each star. For example, setting this property to 1, means that there will be 1 star for every point in the view. A greater number means less stars within the view.
     Defaults to 10000.0.
     */
    @IBInspectable public var numberOfPointsForStar: CGFloat = 10000.0 {
        didSet {
            reload()
        }
    }
    
    /** The size of the stars in points. Defaults to 5.0.
    */
    @IBInspectable public var starSize: CGFloat = 5.0 {
        didSet {
            reload()
        }
    }
    
    /** The color of the stars. Defaults to white.
    */
    @IBInspectable public var starColor = UIColor.whiteColor()
    
    /** The minimum percent by which the stars' size could be changed. For example, a value of 50.0 means that the minimum size of a star will be 50% of the `starSize` property.
     Defaults to 50.0.
     - SeeAlso: `starSize`
    */
    @IBInspectable public var starSizeMinRandomizer = 50.0 {
        didSet {
            let max = starSizeMaxRandomizer < starSizeMinRandomizer ? starSizeMinRandomizer + 1 : starSizeMaxRandomizer
            starSizeProductRandomizerRange = UInt32(starSizeMinRandomizer)...UInt32(max)
        }
    }
    
    /** The maximum percent by which the stars' size could be changed. For example, a value of 150.0 means that the maximum size of a star will be 150% of the `starSize` property.
     Defaults to 150.0.
     - SeeAlso: `starSize`
     */
    @IBInspectable public var starSizeMaxRandomizer = 150.0 {
        didSet {
            let min = starSizeMinRandomizer > starSizeMaxRandomizer ? starSizeMaxRandomizer - 1 : starSizeMinRandomizer
            starSizeProductRandomizerRange = UInt32(min)...UInt32(starSizeMaxRandomizer)
        }
    }
    
    private var starSizeProductRandomizerRange = UInt32(50)...UInt32(150) {
        didSet {
            reload()
        }
    }
    
    /** The stars are drawn with a smaller opacity at the bottom of the view than at the top. This property sets the minimum opacity for the lower stars. Note that the stars at the top of the view will always have an opacity of 1.0. Defaults to 0.5.
    */
    @IBInspectable public var minStarOpacity: Float = 0.5 {
        didSet {
            reload()
        }
    }
    
    /** The intensity of the glowing of the stars, from 0 to 1. If set to 0, the stars will not glow. Defaults to 0.5.
    */
    @IBInspectable public var glowingIntensity: Float = 0.5 {
        didSet {
            reload()
        }
    }
    
    /** The duration in seconds at which the stars glow. Defaults to 2.0.
    */
    @IBInspectable public var glowingDuration = 2.0 {
        didSet {
            reload()
        }
    }
    
    /** Type of stars.
    */
    public enum starTypes {
        /** Round stars.
         */
        case Round
        
        /** Smooth star with 5 branches.
        */
        case FiveBranchStar
    }
    
    /** The type of stars to be drawn. Defaults to `.Round`.
    */
    @IBInspectable public var starType: starTypes = .Round {
        didSet {
            reload()
        }
    }
    
    private let starLayer = CAShapeLayer()
    
    // MARK: Initializers
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    private func commonInit() {
        layer.addSublayer(starLayer)
    }
    
    // MARK: Public Functions
    
    /** Reloads the NightView instance, redrawing all of the stars.
    */
    public func reload() {
        setNeedsDisplay()
        layer.displayIfNeeded()
    }
    
    // MARK: Drawings
    
    override public func layoutSubviews() {
        super.layoutSubviews()
        reload()
    }
    
    override public func drawRect(rect: CGRect) {
        starLayer.frame = bounds
        drawStars()
    }
    
    private func drawStars() {
        removeSublayersFromLayer(starLayer)
        
        let numberOfStars = numberOfStarsForSize(bounds.size)
        for _ in 0..<numberOfStars {
            let layer = drawStarWithFrame(randomStarFrameInFrame(bounds), boundingFrame: bounds)
            if glowingIntensity > 0.0 {
                addGlowingAnimationToStar(layer)
            }
        }
    }
    
    private func drawStarWithFrame(frame: CGRect, boundingFrame: CGRect?) -> CAShapeLayer {
        let layer = CAShapeLayer()
        layer.frame = frame
        layer.fillColor = starColor.CGColor
        layer.path = pathForStarWithType(starType, frame: frame).CGPath
        layer.backgroundColor = UIColor.clearColor().CGColor
        
        if let boundingFrame = boundingFrame {
            layer.opacity = max((Float(boundingFrame.height - frame.origin.y) / 100.0) / (Float(boundingFrame.height) / 100.0), minStarOpacity)
        }
        
        starLayer.addSublayer(layer)
        return layer
    }
    
    // MARK: Paths
    
    private func pathForStarWithType(starType: starTypes, frame: CGRect) -> UIBezierPath {
        let width = frame.width
        
        switch starType {
        case .FiveBranchStar:
            let starPath = UIBezierPath()
            starPath.moveToPoint(CGPoint(x: width / (1000 / 500), y: width / (1000 / 50)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 676.34), y: width / (1000 / 307.29)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 975.53), y: width / (1000 / 395.49)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 785.32), y: width / (1000 / 642.71)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 793.89), y: width / (1000 / 954.51)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 500), y: width / (1000 / 850)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 206.11), y: width / (1000 / 954.51)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 214.68), y: width / (1000 / 642.71)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 24.47), y: width / (1000 / 395.49)))
            starPath.addLineToPoint(CGPoint(x: width / (1000 / 323.66), y: width / (1000 / 307.29)))
            starPath.closePath()
            return starPath
            
        default:
            return UIBezierPath(ovalInRect: CGRectMake(0.0, 0.0, frame.width, frame.height))
        }
    }
    
    // MARK: Animations
    
    private func addGlowingAnimationToStar(layer: CAShapeLayer) {
        let animation = CABasicAnimation(keyPath: "opacity")
        animation.duration = glowingDuration
        animation.fromValue = layer.opacity
        animation.beginTime = CACurrentMediaTime() + Double(arc4random_uniform(UInt32(glowingDuration) * 1000)) / 1000
        animation.toValue = NSNumber(float: layer.opacity - layer.opacity * glowingIntensity)
        animation.autoreverses = true
        animation.repeatCount = .infinity
        layer.addAnimation(animation, forKey: "opacity")
    }
    
    // MARK: Generators
    
    private func randomStarFrameInFrame(frame: CGRect) -> CGRect {
        let size = randomStarSizeForSize(starSize)
        
        let minX = frame.origin.x
        let maxX = frame.origin.x + frame.size.width
        let minY = frame.origin.y
        let maxY = frame.origin.y + frame.size.height
        
        let randomX = CGFloat(arc4random_uniform(UInt32(maxX - minX))) + minX
        let randomY = CGFloat(arc4random_uniform(UInt32(maxY - minY))) + minY
        
        let starFrame = CGRectMake(randomX, randomY, size.width, size.height)
        
        return starFrame
    }
    
    private func randomStarSizeForSize(starSize: CGFloat) -> CGSize {
        let randomStarSizeProduct = CGFloat(arc4random_uniform(starSizeProductRandomizerRange.last! - starSizeProductRandomizerRange.first!) + starSizeProductRandomizerRange.first!) / 100.0;
        let randomSize = starSize * randomStarSizeProduct

        return CGSize(width: randomSize, height: randomSize)
    }
    
    private func numberOfStarsForSize(size: CGSize) -> Int {
        let area = size.width * size.height
        
        return Int(area / numberOfPointsForStar)
    }
    
    private func removeSublayersFromLayer(layer: CAShapeLayer) {
        if let sublayers = layer.sublayers {
            for sublayer in sublayers {
                sublayer.removeFromSuperlayer()
            }
        }
    }
}
